Stash enq_flags when marking for direct dispatch

Decave opened the pull request:
We're currently not remembering enq flags during direct dispatch. Let's
record them in case someone wants to pass e.g. SCX_ENQ_PREEMPT from
ops.select_cpu().

Let's also reset ddsq_id and ddsq_enq_flags before calling
dispatch_enqueue() to ensure there's no races with the task being
consumed from another core.

htejun commented on include/linux/sched/ext.h:
```diff
@@ -690,6 +689,7 @@ struct sched_ext_entity {
 	u64			core_sched_at;	/* see scx_prio_less() */
 #endif
 	u64			ddsq_id;
+	u64			ddsq_enq_flags;
```
`ddsq_id` can be read as "direct dispatch q ID" but I'm not sure how to read `ddsq_enq_flags`. Maybe they should be `ddsp_dsq_id` (or `ddsp_qid`) and `ddsp_enq_flags`?

htejun commented on kernel/sched/ext.c:
```diff
@@ -847,14 +845,11 @@ static void direct_dispatch(struct task_struct *p, u64 enq_flags)
-		p->scx.flags &= ~SCX_TASK_DDSP_PRIQ;
-	}
-
+	enq_flags |= p->scx.ddsq_enq_flags | SCX_ENQ_CLEAR_OPSS;
```
Can we update `scx_bpf_dispatch()` documentation explaining what happens to `enq_flags` when direct dispatching from `.select_cpu()`? This isn't immediately obvious.

Decave replied:
Will do. FWIW, I think we should eventually consider updating scx_bpf_dispatch*() to not take enq_flags, and to instead take some flags that we define as part of the sched_ext API. Passing enq flags feels like it leaks implementation details of the core scheduler.

htejun replied:
Oh, left the comment in the wrong thread. Here it goes again: Yeah, we can drop the lower 32 bits of enq flags but the currently shared flags are `WAKEUP` and `HEAD`, both of which are well defined and should be passed to `ops.enqueue()`. We can redefine them so that we don't pass them through verbatim but I don't think it'd make whole lot of difference.

htejun commented on tools/testing/selftests/scx/select_cpu_vtime.c:
```diff
@@ -0,0 +1,50 @@
+	for (i = 0; i < NUM_CHILDREN; i++) {
+		SCX_EQ(waitpid(pids[i], &status, 0), pids[i]);
+		SCX_EQ(status, 0);
+	}
```
I have a hard time understanding what the test cases are testing. This looks like a basic vtime scheduler and we're forking and waiting a number of children. It'd be useful if each test case states what it aims to test and how it's doing that.

Decave replied:
Sure, I'll document this more thoroughly. FWIW, the purposes of these testcases will become much more clear when I'm done with the test refactoring. The API for adding testcases will require the user to document what they're testing.

htejun replied:
Yeah, maybe we drop the lower 32 bits which are shared with the core flags. That said, the shared flags we expose are `WAKEUP` and `HEAD`, both of which are well-defined and should be passed to enqueue.

htejun reviewed (approved):
Generally LGTM. Minor comments.

Decave commented:
Addressed feedback and also added https://github.com/sched-ext/sched_ext/pull/115/commits/94cfcaced95b8dcc65b0c19e490a27e627a74112, which fixes direct dispatch from ops.select_cpu() for non-builtin DSQs

htejun commented on kernel/sched/ext.c:
```diff
@@ -680,6 +680,7 @@ static void dispatch_enqueue(struct scx_dispatch_q *dsq, struct task_struct *p,
 	}
 	dsq->nr++;
 	p->scx.dsq = dsq;
+	p->scx.ddsq_id = SCX_DSQ_INVALID;
```
Please explain that we need ddsq_id cleared here because the stored direct dispatch verdict may be overridden in the enqueue path.
